File: go/internal/sol/p0001/p0001.go
package p0001

import "strconv"

func compute(limit uint) string {
	var f = func(x uint) uint {
		var tmp = (limit - 1) / x
		return (1 + tmp) * tmp / 2 * x
	}

	return strconv.FormatUint(uint64(f(3)+f(5)-f(15)), 10)
}

func Solve() string {
	return compute(1_000)
}

File: haskell/src/Sol/P0001.hs
module Sol.P0001 (compute, solve) where

compute :: Int -> String
compute limit =
    show $ sumOfMultiples 3 + sumOfMultiples 5 - sumOfMultiples 15
  where
    upper = pred limit

    sumOfMultiples :: Int -> Int
    sumOfMultiples n =
        div ((n + (upper - mod upper n)) * div upper n) 2

solve :: String
solve = compute 1000
